br/pkg/lightning/importer/table_import.go

```diff
@@ -1176,3 +1202,303 @@ func (tr *TableImporter) analyzeTable(ctx context.Context, g glue.SQLExecutor) e
 	task.End(zap.ErrorLevel, err)
 	return err
 }
+
+func (tr *TableImporter) addIndexes(ctx context.Context, db *sql.DB) (retErr error) {
+	const progressStep = "add-index"
+	task := tr.logger.Begin(zap.InfoLevel, "add indexes")
+	defer func() {
+		task.End(zap.ErrorLevel, retErr)
+		if retErr == nil {
+			web.BroadcastTableProgress(tr.tableName, progressStep, 1)
+		}
+	}()
+
+	tblInfo := tr.tableInfo
+	tableName := tr.tableName
+
+	singleSQL, multiSQLs := buildAddIndexSQL(tableName, tblInfo.Core, tblInfo.Desired)
+	if len(multiSQLs) == 0 {
+		return nil
+	}
+
+	logger := log.FromContext(ctx).With(zap.String("table", tableName))
+
+	// Try to add all indexes in one statement.
+	err := tr.executeDDL(ctx, db, singleSQL, len(multiSQLs), func(p float64) {
+		web.BroadcastTableProgress(tableName, progressStep, p)
+		logger.Info("add index progress", zap.String("progress", fmt.Sprintf("%.1f%%", p*100)))
+	})
+	if err == nil {
+		return nil
+	}
+	if !isDupKeyError(err) || len(multiSQLs) == 1 {
+		return err
+	}
+	logger.Warn("cannot add all indexes in one statement, try to add them one by one", zap.Strings("sqls", multiSQLs), zap.Error(err))
+
+	baseProgress := float64(0)
+	for _, ddl := range multiSQLs {
+		err := tr.executeDDL(ctx, db, ddl, 1, func(p float64) {
+			progress := baseProgress + p/float64(len(multiSQLs))
+			web.BroadcastTableProgress(tableName, progressStep, baseProgress+p/float64(len(multiSQLs)))
+			logger.Info("add index progress", zap.String("progress", fmt.Sprintf("%.1f%%", progress*100)))
+		})
+		if err != nil && !isDupKeyError(err) {
+			return err
+		}
+		baseProgress += 1.0 / float64(len(multiSQLs))
+		web.BroadcastTableProgress(tableName, progressStep, baseProgress)
+	}
+	return nil
+}
+
+func (tr *TableImporter) executeDDL(
+	ctx context.Context,
+	db *sql.DB,
+	ddl string,
+	indexCount int,
```

Comment:
From the name `executeDDL`, it is a generic function to execute a DDL statement, but the `indexCount` parameter is only used to calculate the progress, and only specific to calculating the adding index progress.  This is a bit strange. 

What about change the parameter of `updateProgress` from a float64 to a ddlStatus struct.  Then in this function it can: 1) calculate the progress; 2) update the progress .  

Or we can just pass in a interface called `ProgressUpdater`, which provides two methods: 1) `CalculateProgress(status *ddlStatus) float64` and 2) `UpdateProgress(progress float64)` .  In this way, it decouples the DDL executing and the progress updating, and we can make `indexCount` as a state in a `ProgressUpdater` object.

Reply:
Good idea! I will try to improve it.

Reply:
https://github.com/pingcap/tidb/pull/42136/commits/b2660b374168fc10dbae5d5811998f2dff44e456

Reply:
Sorry, the correct commit is https://github.com/pingcap/tidb/pull/42136/commits/4457efb0bdebc9510bd458e5c63ca4ca52cb5332.